### PR TITLE
Hijack triggers

### PR DESCRIPTION
**Pull request by GANDAMU**
hi
team
Hey team.
TEAM.
HEY TEAM.

I think...this might actually give us the piece we need for interactivity? Hopefully this is the branch that gets us to meaningful progress on the Pong game?

The "test" sequence:

1. `oscsend localhost 5006 /trigger/left_nail` - See that left_nail triggers "braindead"
2. `oscsend localhost 5006 /rosary/add_effect/name/trigger_hijack_demo` - When "braindead" ends, do this and see the demo effect
3. `oscsend localhost 5006 /trigger/left_nail` - See that the left_nail now reverses the "looper" effect's direction when fired when the red looper is on the left-side whitened bead
4. `oscsend localhost 5006 /rosary/clear_effects` - Remove the demo effect
5. `oscsend localhost 5006 /trigger/left_nail` - See that left_nail once again triggers "braindead"

### CODE REVIEW

File: python/mp/rosary.py [hunk 1]
```diff
@@ -535,22 +548,38 @@ def mainloop(self, *args, **kwargs):
...
+            # effect's hijacked trigger fires
+            hijacked_effect, hijacked_method = hijacked[-1]
+
+            print("Trigger {} hijacked by {}".format(trigger_name,
+                                                     hijacked_effect))
+            hijacked_method()
+
+        else:
+
+            requested_trigger = self.trigger_registry.get(trigger_name)
```
Comment (jdblair) [inline python/mp/rosary.py:570]: this is the original behavior, right?

**Comment by jdblair** [comment]
I like this, I think its awesome. I don't have specific comments yet, but it gave me an idea how to write the next feature we need: "idle" behavior. E.g., what is the rosary doing when its just sitting there? I was struggling with how to design this, but now I see we can just write an effect that handles it, and processes input to decide what to do, and start other effects.

File: python/mp/effects/__init__.py [hunk 2]
```diff
@@ -23,5 +23,6 @@
     'looper',
     'wheel',
     'roulette',
-    'random_fill'
+    'random_fill',
+    'trigger_hijack_demo'
 ]
```
Comment (jdblair) [inline python/mp/effects/__init__.py]: If you add your new effect in the middle of the list then we won't have conflicts when merging multiple new effects.

File: python/mp/rosary.py [hunk 3]
```diff
@@ -535,22 +548,38 @@ def mainloop(self, *args, **kwargs):
...
+        # See if any currently running effects have hijacked this trigger
+        hijacked = self.trigger_hijacks.get(trigger_name)
+
+        if hijacked is not None and len(hijacked) > 0:
+
+            # I thought it'd be cool if the hijacks acted like a stack -
+            # if many effects hijack the same trigger, then only the newest
+            # effect's hijacked trigger fires
+            hijacked_effect, hijacked_method = hijacked[-1]
+
```
Comment (jdblair) [inline python/mp/rosary.py:563]: Does this mean I can do this?
1) Effect A hijacks the triggers, and at some point starts effect B.
2) Effect B hijacks the triggers and does stuff.
3) Effect B exits (sets self.finished=true), Effect A processes triggers again.

**Comment by jdblair** [comment]
+1. Merge it. My only objection is "hijacking" is a weird term to use, but I'm getting used to it.

File: python/mp/rosary.py (hunk 3 above)
Comment (GANDAMU) [reply python/mp/rosary.py:563]: That is the intention.

Although I've only tested 1 "layer" so far (if you consider what you're proposing as having 2 "layers")

File: python/mp/rosary.py (hunk 1 above)
Comment (GANDAMU) [reply python/mp/rosary.py:570]: Yep, I just shoved the entirety of the old method into the else.

File: python/mp/effects/__init__.py (hunk 2 above)
Comment (GANDAMU) [reply python/mp/effects/__init__.py]: If we had more time I'd ask "lordy is there no way to just import all the classes and be done with it?"

**Review by GANDAMU — COMMENTED** [review]
Okay, gonna fix that conflict and merge it in.

**Comment by GANDAMU** [comment]
Also I love the term "hijacking" here. It feels so COOL AND HACKERY.

**Pull request merged**